doc/developer_guide.md

```diff
@@ -0,0 +1,418 @@
+modules (netmod). We currently support `ch3:nemesis:tcp` and
+`ch3:nemesis:ofi`.
+
+### ch4
+
+Ch4 is where currently active research and development take place. Many new
+features, e.g. per-VCI threading, GPU IPC, partitioned communication, etc. are
+only available in Ch4.
+
+Ch4 designs an additional ADI-like interface, commonly referred as ch4 API or
+shm/netmod API. With most MPID functions, ch4-layer will check whether the
+communication is local (can be carried out using shared memory) and calls into
+either shm API or netmod API. It is possible to disable shm entirely.
+
+The framework for ch4 API involves large amount of boilerplates due to the
+need allow both fully inlined build and nonlined build using function tables.
+We use script to generate most of these API files.
+
+Reference:
+* [Notes on ch4 api autogen](notes/ch4_api.md)
+* [Notes on ch4 inline mechanism](notes/ch4_inline.txt)
+* [Notes on ch4 namespace convention](notes/ch4_namespace.txt)
+* [ch4_api.txt](../src/mpid/ch4/ch4_api.txt)
+* [request.md](notes/request.md)
+
+#### Active Message
+
+Similar to MPIR, ch4 provide a default/fallback implementation based on active
+messages. Ch3 implementation is largely an active message implementation. It
+is easier to use an example to explain what is active message and how it
+works.
+
+MPID_Send will call either `MPIDI_SHM_mpi_isend` or `MPIDI_NM_mpi_isend` based
+on locality, which may call back to `MPIDIG_mpi_isend`. `MPIDIG_mpi_isend` will
+send the message along with am header to target process using
+`MPIDI_{SHM,NM}_am_isend`. It creates an MPIR_Request object and return to user.
+
+During progress, the receiving process receives the message (inside shm
+progress or netmod progress), it will check the am header, and call a
+registered handler function to process the message. The handler functions are
+essential pieces in a active message protocol. In the pt2pt eager send case,
+it will check posted receiver buffer list and either copy the message data to
+the receiver buffer (and completes the receive request) or enqueuing the
+message to an unexpected queue.
+
+As a counter part, `MPID_Recv` will at some point call `MPIDIG_mpi_irecv`, which
+will check the unexpected queue, and either copy the data over (and completes
+the receive buffer) or enqueue the receive request to a posted queue.
+
+There are complications, where we may implement different protocols such as
+rendezvous protocol or IPC or RDMA, as well as various RMA operations.
+Different protocols are all implemented by handler functions which in turn may
+send additional active messages with another handler functions to carry on the
+protocols.
+
+Reference:
+* src/mpid/ch4/generic/am/mpidig_am_send.h
+* src/mpid/ch4/generic/am/mpidig_am_recv.h
+* src/mpid/ch4/src/ch4r_callbacks.h
+* src/mpid/ch4/src/ch4r_recvq.h
+* src/mpid/ch4/src/ch4r_rma.h
+* src/mpid/ch4/src/ch4r_rma_target_callbacks.h
+* src/mpid/ch4/src/ch4r_rma_origin_callbacks.h
```

Comment:
Should these link to the appropriate files in the repo?

Reply:
Fixed